clippy_lints/src/recursive_format_impl.rs

```diff
@@ -0,0 +1,269 @@
+        if let Some(format_trait_impl) = is_format_trait_impl(cx, item) {
+            self.format_trait_impl = Some(format_trait_impl);
+        }
+    }
+
+    fn check_item_post(&mut self, cx: &LateContext<'_>, item: &Item<'_>) {
+        // Assume no nested Impl of Debug and Display within eachother
+        if is_format_trait_impl(cx, item).is_some() {
+            self.format_trait_impl = None;
+            self.self_hir_id = None;
+        }
+    }
+
+    fn check_impl_item(&mut self, cx: &LateContext<'_>, impl_item: &ImplItem<'_>) {
+        if_chain! {
+            // If we are in Display or Debug impl, then get hir_id for self in method impl - i.e. fmt(&self)
+            if self.format_trait_impl.is_some();
+            if let ImplItemKind::Fn(.., body_id) = &impl_item.kind;
+            let body = cx.tcx.hir().body(*body_id);
+            if !body.params.is_empty();
+            then {
+                let self_param = &body.params[0];
+                self.self_hir_id = Some(self_param.pat.hir_id);
+            }
+        }
+    }
+
+    fn check_expr(&mut self, cx: &LateContext<'_>, expr: &Expr<'_>) {
+        if let Some(self_hir_id) = self.self_hir_id {
+            match self.format_trait_impl {
+                Some(ImplTrait::Display) => {
+                    check_to_string_in_display(cx, expr, self_hir_id);
+                    check_self_in_format_args(cx, expr, self_hir_id, ImplTrait::Display);
+                },
+                Some(ImplTrait::Debug) => {
+                    check_self_in_format_args(cx, expr, self_hir_id, ImplTrait::Debug);
+                },
+                None => {},
+            }
+        }
+    }
+}
+
+fn check_to_string_in_display(cx: &LateContext<'_>, expr: &Expr<'_>, self_hir_id: HirId) {
+    if_chain! {
+        // Get the hir_id of the object we are calling the method on
+        if let ExprKind::MethodCall(path, _, [ref self_arg, ..], _) = expr.kind;
+        // Is the method to_string() ?
+        if path.ident.name == sym!(to_string);
+        // Is the method a part of the ToString trait? (i.e. not to_string() implemented
+        // separately)
+        if let Some(expr_def_id) = cx.typeck_results().type_dependent_def_id(expr.hir_id);
+        if is_diag_trait_item(cx, expr_def_id, sym::ToString);
+        // Is the method is called on self
+        if path_to_local_id(self_arg, self_hir_id);
+        then {
+            span_lint(
+                cx,
+                RECURSIVE_FORMAT_IMPL,
+                expr.span,
+                "using `to_string` in `fmt::Display` implementation might lead to infinite recursion",
+            );
+        }
+    }
+}
+
+fn check_self_in_format_args(cx: &LateContext<'_>, expr: &Expr<'_>, self_hir_id: HirId, impl_trait: ImplTrait) {
+    // Check each arg in format calls - do we ever use Display on self (directly or via deref)?
+    if_chain! {
+            if let Some(format_args) = FormatArgsExpn::parse(cx, expr);
```

Comment:
We have some new macro utils that should be used here. First call `root_macro_call_first_node` and then `FormatArgsExpn::parse` (this will avoid some calls to `FormatArgsExpn::parse` which can be expensive). The output of `root_macro_call_first_node` can be used instead of `outermost_expn_data`.

Reply:
Which HIR node / expr should be checked here?

It works if I just use:
```
        if let Some(macro_call) = root_macro_call(expr.span);
```

But trying to add the first node check with the `expr` results in it always being None.

Reply:
Okay, I think the problem is that `format_args!` is not the root macro call. It is nested within a format macro. So you should check the outer macro with `root_macro_call_first_node` and `is_format_macro` and then use `FormatArgsExpn::find_nested` to find the nested `format_args!` call.

Reply:
Nice, this also has the benefit of providing the outer macro name for the error message :slightly_smiling_face: